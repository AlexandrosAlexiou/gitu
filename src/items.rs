use crate::diff;
use crate::theme;
use crate::Config;
use ansi_to_tui::IntoText;
use diff::Delta;
use diff::Hunk;
use ratatui::style::Style;
use ratatui::text::Text;
use std::borrow::Cow;
use std::io::Write;
use std::iter;
use std::process::Command;
use std::process::Stdio;

#[derive(Default, Clone, Debug, PartialEq, Eq, Hash)]
pub(crate) struct Item {
    pub(crate) id: Cow<'static, str>,
    pub(crate) display: Text<'static>,
    pub(crate) section: bool,
    pub(crate) depth: usize,
    pub(crate) unselectable: bool,
    pub(crate) target_data: Option<TargetData>,
}

#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub(crate) enum TargetData {
    Ref(String),
    File(String),
    Delta(Delta),
    Hunk(Hunk),
}

pub(crate) fn create_diff_items<'a>(
    config: &'a Config,
    diff: &'a diff::Diff,
    depth: &'a usize,
) -> impl Iterator<Item = Item> + 'a {
    diff.deltas.iter().flat_map(|delta| {
        let target_data = TargetData::Delta(delta.clone());

        iter::once(Item {
            id: delta.file_header.to_string().into(),
            display: Text::styled(
                if delta.old_file == delta.new_file {
                    delta.new_file.clone()
                } else {
                    format!("{} -> {}", delta.old_file, delta.new_file)
                },
                Style::new().fg(theme::CURRENT_THEME.file),
            ),
            section: true,
            depth: *depth,
            target_data: Some(target_data),
            ..Default::default()
        })
        .chain(
            delta
                .hunks
                .iter()
                .flat_map(|hunk| create_hunk_items(config, hunk, *depth)),
        )
    })
}

fn create_hunk_items(config: &Config, hunk: &Hunk, depth: usize) -> impl Iterator<Item = Item> {
    let target_data = TargetData::Hunk(hunk.clone());

    iter::once(Item {
        id: hunk.format_patch().into(),
        display: Text::styled(
            hunk.header(),
            Style::new().fg(theme::CURRENT_THEME.hunk_header),
        ),
        section: true,
        depth: depth + 1,
        target_data: Some(target_data),
        ..Default::default()
    })
    .chain([{
        Item {
            display: format_diff_hunk(config, hunk)
                .into_text()
                .expect("Error creating hunk text"),
            unselectable: true,
            depth: depth + 2,
            target_data: None,
            ..Default::default()
        }
    }])
}

fn format_diff_hunk(config: &Config, hunk: &Hunk) -> String {
    if config.use_delta {
        let content = format!("{}\n{}", hunk.header(), hunk.content);
        {
            let input = content.as_bytes();
            let cmd: &[&str] = &[
                "delta",
                "--color-only",
                &format!("-w {}", crossterm::terminal::size().unwrap().0),
            ];
            let mut command = Command::new(cmd[0])
                .args(&cmd[1..])
                .stdin(Stdio::piped())
                .stdout(Stdio::piped())
                .spawn()
                .expect("Couldn't spawn process");

            command
                .stdin
                .take()
                .unwrap_or_else(|| panic!("No stdin for {:?} process", cmd))
                .write_all(input)
                .expect("Couldn't write to process stdin");
            let output = command
                .wait_with_output()
                .unwrap_or_else(|_| panic!("Error writing {:?} output", cmd));

            String::from_utf8(output.stdout).unwrap()
        }
        .lines()
        .skip(1) // Header is already shown
        .collect::<Vec<_>>()
        .join("\n")
    } else {
        hunk.content.clone()
    }
}

pub(crate) fn create_log_items(log: &str) -> impl Iterator<Item = Item> + '_ {
    log.lines().map(|log_line| {
        let target_data = TargetData::Ref(strip_ansi_escapes::strip_str(
            log_line
                .split_whitespace()
                .next()
                .expect("Error extracting ref"),
        ));

        Item {
            id: log_line.to_string().into(),
            display: log_line
                .to_string()
                .into_text()
                .expect("Error creating log text"),
            depth: 1,
            target_data: Some(target_data),
            ..Default::default()
        }
    })
}
