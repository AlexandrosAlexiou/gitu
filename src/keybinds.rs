use crossterm::event::{self, KeyCode, KeyModifiers};

use KeyCode::*;
use Op::*;
use SubmenuOp::*;
use TargetOp::*;

pub(crate) struct Keybind {
    pub submenu: SubmenuOp,
    pub mods: KeyModifiers,
    pub key: KeyCode,
    pub op: Op,
}

impl Keybind {
    const fn nomod(submenu: SubmenuOp, key: KeyCode, op: Op) -> Self {
        Self {
            submenu,
            mods: KeyModifiers::NONE,
            key,
            op,
        }
    }

    const fn ctrl(submenu: SubmenuOp, key: KeyCode, op: Op) -> Self {
        Self {
            submenu,
            mods: KeyModifiers::CONTROL,
            key,
            op,
        }
    }

    const fn shift(submenu: SubmenuOp, key: KeyCode, op: Op) -> Self {
        Self {
            submenu,
            mods: KeyModifiers::SHIFT,
            key,
            op,
        }
    }

    pub(crate) fn format_key(&self) -> String {
        let modifiers = self
            .mods
            .iter_names()
            .map(|(name, _)| match name {
                "CONTROL" => "C-",
                "SHIFT" => "",
                _ => unimplemented!("format_key mod {}", name),
            })
            .collect::<String>();

        modifiers
            + &match self.key {
                KeyCode::Enter => "ret".to_string(),
                KeyCode::Left => "←".to_string(),
                KeyCode::Right => "→".to_string(),
                KeyCode::Up => "↑".to_string(),
                KeyCode::Down => "↓".to_string(),
                KeyCode::Tab => "tab".to_string(),
                KeyCode::Delete => "del".to_string(),
                KeyCode::Insert => "ins".to_string(),
                KeyCode::F(n) => format!("F{}", n),
                KeyCode::Char(c) => if self.mods.contains(KeyModifiers::SHIFT) {
                    c.to_ascii_uppercase()
                } else {
                    c
                }
                .to_string(),
                KeyCode::Esc => "esc".to_string(),
                _ => "???".to_string(),
            }
    }
}

pub(crate) const KEYBINDS: &[Keybind] = &[
    // Generic
    Keybind::nomod(Any, Char('q'), Quit),
    Keybind::nomod(Any, Esc, Quit),
    Keybind::nomod(None, Char('g'), Refresh),
    Keybind::nomod(None, Tab, ToggleSection),
    // Navigation
    Keybind::nomod(None, Char('k'), SelectPrevious),
    Keybind::nomod(None, Char('p'), SelectPrevious),
    Keybind::nomod(None, KeyCode::Up, SelectPrevious),
    Keybind::nomod(None, Char('j'), SelectNext),
    Keybind::nomod(None, Char('n'), SelectNext),
    Keybind::nomod(None, KeyCode::Down, SelectNext),
    Keybind::ctrl(None, Char('u'), HalfPageUp),
    Keybind::ctrl(None, Char('d'), HalfPageDown),
    // Help
    Keybind::nomod(None, Char('h'), Submenu(Help)),
    // Branch
    Keybind::nomod(None, Char('b'), Submenu(Branch)),
    Keybind::nomod(Branch, Char('b'), Target(Checkout)),
    // Commit
    Keybind::nomod(None, Char('c'), Submenu(SubmenuOp::Commit)),
    Keybind::nomod(SubmenuOp::Commit, Char('c'), Op::Commit),
    Keybind::nomod(SubmenuOp::Commit, Char('a'), CommitAmend),
    Keybind::nomod(SubmenuOp::Commit, Char('f'), Target(CommitFixup)),
    // Fetch
    Keybind::nomod(None, Char('f'), Submenu(Fetch)),
    Keybind::nomod(Fetch, Char('a'), FetchAll),
    // Log
    Keybind::nomod(None, Char('l'), Submenu(Log)),
    Keybind::nomod(Log, Char('l'), LogCurrent),
    // Pull
    Keybind::shift(None, Char('F'), Submenu(Pull)),
    Keybind::nomod(Pull, Char('p'), PullRemote),
    // Push
    Keybind::shift(None, Char('P'), Submenu(Push)),
    Keybind::nomod(Push, Char('p'), PushRemote),
    // Rebase
    Keybind::nomod(None, Char('r'), Submenu(Rebase)),
    Keybind::nomod(Rebase, Char('i'), Target(RebaseInteractive)),
    Keybind::nomod(Rebase, Char('a'), RebaseAbort),
    Keybind::nomod(Rebase, Char('c'), RebaseContinue),
    Keybind::nomod(Rebase, Char('f'), Target(RebaseAutosquash)),
    // Show refs
    Keybind::nomod(None, Char('y'), ShowRefs),
    // Discard
    Keybind::shift(None, Char('K'), Submenu(SubmenuOp::Discard)),
    Keybind::nomod(SubmenuOp::Discard, Char('y'), Target(TargetOp::Discard)),
    Keybind::nomod(SubmenuOp::Discard, Char('n'), Quit),
    // Target actions
    Keybind::nomod(None, Enter, Target(Show)),
    Keybind::nomod(None, Char('s'), Target(Stage)),
    Keybind::nomod(None, Char('u'), Target(Unstage)),
];

#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub(crate) enum Op {
    Quit,
    Refresh,
    SelectPrevious,
    SelectNext,
    ToggleSection,
    HalfPageUp,
    HalfPageDown,
    PushRemote,
    PullRemote,
    Submenu(SubmenuOp),
    Commit,
    CommitAmend,
    FetchAll,
    LogCurrent,
    RebaseAbort,
    RebaseContinue,
    ShowRefs,
    Target(TargetOp),
}

#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub(crate) enum SubmenuOp {
    Any,
    None,
    Branch,
    Commit,
    Discard,
    Fetch,
    Help,
    Log,
    Pull,
    Push,
    Rebase,
}

#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub(crate) enum TargetOp {
    Checkout,
    CommitFixup,
    Show,
    Stage,
    Unstage,
    RebaseAutosquash,
    RebaseInteractive,
    Discard,
}

impl TargetOp {
    // TODO find a better way to always keep this up to date with TargetOp contents
    pub(crate) fn list_all() -> impl Iterator<Item = &'static TargetOp> {
        [
            &TargetOp::Checkout,
            &TargetOp::CommitFixup,
            &TargetOp::Show,
            &TargetOp::Stage,
            &TargetOp::Unstage,
            &TargetOp::RebaseAutosquash,
            &TargetOp::RebaseInteractive,
            &TargetOp::Discard,
        ]
        .into_iter()
    }
}

pub(crate) fn op_of_key_event(pending: SubmenuOp, key: event::KeyEvent) -> Option<Op> {
    KEYBINDS
        .iter()
        .find(|keybind| {
            (keybind.submenu, keybind.mods, keybind.key) == (pending, key.modifiers, key.code)
                || (keybind.submenu, keybind.mods, keybind.key) == (Any, key.modifiers, key.code)
        })
        .map(|keybind| keybind.op)
}

pub(crate) fn list(pending: &SubmenuOp) -> impl Iterator<Item = &Keybind> {
    let expected = if pending == &Help { None } else { *pending };

    KEYBINDS
        .iter()
        .filter(move |keybind| keybind.submenu == expected)
}
