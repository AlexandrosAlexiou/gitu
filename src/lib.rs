pub mod cli;
mod git;
mod git2_opts;
mod items;
mod keybinds;
mod screen;
mod theme;
mod ui;

use crossterm::event::{self, Event, KeyEventKind};
use git2::Repository;
use items::{Item, TargetData};
use itertools::Itertools;
use keybinds::{Op, SubmenuOp, TargetOp};
use ratatui::{prelude::*, Terminal};
use screen::Screen;
use std::{
    borrow::Cow,
    error::Error,
    iter,
    path::PathBuf,
    process::{Command, Stdio},
    rc::Rc,
};

type Res<T> = Result<T, Box<dyn Error>>;

#[derive(Clone)]
struct Config {
    dir: PathBuf,
}

pub(crate) struct CmdMeta {
    pub(crate) args: Cow<'static, str>,
    pub(crate) out: Option<String>,
}

struct State {
    repo: Rc<Repository>,
    config: Config,
    quit: bool,
    screens: Vec<Screen>,
    pending_submenu_op: SubmenuOp,
    pub(crate) cmd_meta: Option<CmdMeta>,
}

impl State {
    fn create(repo: Repository, config: Config, size: Rect, args: cli::Args) -> Res<Self> {
        let repo = Rc::new(repo);
        repo.set_workdir(&config.dir, false)?;

        let screens = match args.command {
            Some(cli::Commands::Show { reference }) => {
                vec![screen::show::create(Rc::clone(&repo), size, reference)?]
            }
            None => vec![screen::status::create(Rc::clone(&repo), &config, size)?],
        };

        Ok(Self {
            repo,
            config,
            quit: args.exit_immediately,
            screens,
            pending_submenu_op: SubmenuOp::None,
            cmd_meta: None,
        })
    }

    fn screen_mut(&mut self) -> &mut Screen {
        self.screens.last_mut().expect("No screen")
    }

    fn screen(&self) -> &Screen {
        self.screens.last().expect("No screen")
    }

    pub(crate) fn run_external_cmd<B: Backend>(
        &mut self,
        terminal: &mut Terminal<B>,
        input: &[u8],
        mut cmd: Command,
    ) -> Res<()> {
        cmd.current_dir(&self.config.dir);

        cmd.stdin(Stdio::piped());
        cmd.stdout(Stdio::piped());
        cmd.stderr(Stdio::piped());

        self.run_cmd(terminal, command_args(&cmd), |_state| {
            let mut child = cmd.spawn()?;

            use std::io::Write;
            child.stdin.take().unwrap().write_all(input)?;

            let out = String::from_utf8(child.wait_with_output()?.stderr.clone())
                .expect("Error turning command output to String");

            Ok(out)
        })?;

        Ok(())
    }

    pub(crate) fn run_cmd<
        B: Backend,
        S: Into<Cow<'static, str>>,
        F: FnMut(&mut Self) -> Res<String>,
    >(
        &mut self,
        terminal: &mut Terminal<B>,
        display: S,
        mut cmd: F,
    ) -> Res<()> {
        self.cmd_meta = Some(CmdMeta {
            args: display.into(),
            out: None,
        });
        terminal.draw(|frame| ui::ui::<B>(frame, &*self))?;

        self.cmd_meta.as_mut().unwrap().out = Some(cmd(self)?);
        self.screen_mut().update()?;

        Ok(())
    }

    pub(crate) fn issue_subscreen_command<B: Backend>(
        &mut self,
        terminal: &mut Terminal<B>,
        mut cmd: Command,
    ) -> Res<()> {
        cmd.current_dir(&self.config.dir);

        cmd.stdin(Stdio::piped());
        let child = cmd.spawn()?;

        let out = child.wait_with_output()?;

        self.cmd_meta = Some(CmdMeta {
            args: command_args(&cmd),
            out: Some(
                String::from_utf8(out.stderr.clone())
                    .expect("Error turning command output to String"),
            ),
        });

        terminal.hide_cursor()?;
        terminal.clear()?;

        self.screen_mut().update()?;

        Ok(())
    }
}

fn command_args(cmd: &Command) -> Cow<'static, str> {
    iter::once(cmd.get_program().to_string_lossy())
        .chain(cmd.get_args().map(|arg| arg.to_string_lossy()))
        .join(" ")
        .into()
}

pub fn run<B: Backend>(args: cli::Args, terminal: &mut Terminal<B>) -> Result<(), Box<dyn Error>> {
    let mut state = State::create(
        Repository::open_from_env()?,
        Config {
            dir: String::from_utf8(
                Command::new("git")
                    .args(["rev-parse", "--show-toplevel"])
                    .output()?
                    .stdout,
            )?
            .trim_end()
            .into(),
        },
        terminal.size()?,
        args,
    )?;
    terminal.draw(|frame| ui::ui::<B>(frame, &state))?;

    while !state.quit {
        // TODO Gather all events, no need to draw for every
        if !event::poll(std::time::Duration::from_millis(u64::MAX))? {
            continue;
        }

        let event = event::read()?;
        update(terminal, &mut state, &[event])?;
    }

    Ok(())
}

pub(crate) fn update<B: Backend>(
    terminal: &mut Terminal<B>,
    state: &mut State,
    events: &[Event],
) -> Res<()> {
    for event in events {
        match *event {
            Event::Resize(w, h) => {
                for screen in state.screens.iter_mut() {
                    screen.size = Rect::new(0, 0, w, h);
                }
            }
            Event::Key(key) => {
                if key.kind == KeyEventKind::Press {
                    state.cmd_meta = None;

                    handle_op(terminal, state, key)?;
                }
            }
            _ => (),
        }
    }

    if let Some(screen) = state.screens.last_mut() {
        screen.clamp_cursor();
        terminal.draw(|frame| ui::ui::<B>(frame, &*state))?;
    }

    Ok(())
}

fn handle_op<B: Backend>(
    terminal: &mut Terminal<B>,
    state: &mut State,
    key: event::KeyEvent,
) -> Res<()> {
    let pending = if state.pending_submenu_op == SubmenuOp::Help {
        SubmenuOp::None
    } else {
        state.pending_submenu_op
    };

    if let Some(op) = keybinds::op_of_key_event(pending, key) {
        use Op::*;
        let was_submenu = state.pending_submenu_op != SubmenuOp::None;
        state.pending_submenu_op = SubmenuOp::None;

        match op {
            Quit => {
                if was_submenu {
                    // Do nothing, already cleared
                } else {
                    state.screens.pop();
                    if let Some(screen) = state.screens.last_mut() {
                        screen.update()?;
                    } else {
                        state.quit = true
                    }
                }
            }
            Refresh => state.screen_mut().update()?,
            ToggleSection => state.screen_mut().toggle_section(),
            SelectPrevious => state.screen_mut().select_previous(),
            SelectNext => state.screen_mut().select_next(),
            HalfPageUp => state.screen_mut().scroll_half_page_up(),
            HalfPageDown => state.screen_mut().scroll_half_page_down(),
            Commit => {
                state.issue_subscreen_command(terminal, git::commit_cmd())?;
            }
            CommitAmend => {
                state.issue_subscreen_command(terminal, git::commit_amend_cmd())?;
            }
            Submenu(op) => state.pending_submenu_op = op,
            LogCurrent => goto_log_screen(Rc::clone(&state.repo), &mut state.screens),
            FetchAll => {
                state.run_external_cmd(terminal, &[], git::fetch_all_cmd())?;
            }
            PullRemote => state.run_external_cmd(terminal, &[], git::pull_cmd())?,
            PushRemote => state.run_cmd(terminal, "git push", |state| {
                git::push_to_matching_remote(&state.repo)
            })?,
            Target(target_op) => {
                if let Some(act) = &state.screen_mut().get_selected_item().target_data.clone() {
                    if let Some(mut closure) = closure_by_target_op(act, &target_op) {
                        closure(terminal, state)?;
                    }
                }
            }
            RebaseAbort => {
                state.run_external_cmd(terminal, &[], git::rebase_abort_cmd())?;
            }
            RebaseContinue => {
                state.run_external_cmd(terminal, &[], git::rebase_continue_cmd())?;
            }
            ShowRefs => goto_refs_screen(&state.config, &mut state.screens),
        }
    }

    Ok(())
}

pub(crate) fn list_target_ops<B: Backend>(
    target: &TargetData,
) -> impl Iterator<Item = &'static TargetOp> + '_ {
    TargetOp::list_all().filter(|target_op| closure_by_target_op::<B>(target, target_op).is_some())
}

type OpClosure<'a, B> = Box<dyn FnMut(&mut Terminal<B>, &mut State) -> Res<()> + 'a>;

/// Retrieves the 'implementation' of a `TargetOp`.
/// These are `Option<OpClosure>`s, so that the mappings
/// can be introspected.
pub(crate) fn closure_by_target_op<'a, B: Backend>(
    target: &'a TargetData,
    target_op: &TargetOp,
) -> Option<OpClosure<'a, B>> {
    use TargetData::*;
    use TargetOp::*;

    match (target_op, target) {
        (Show, Ref(r)) => goto_show_screen(r.clone()),
        (Show, File(u)) => editor(u.clone(), None),
        (Show, Delta(d)) => editor(d.new_file.clone(), None),
        (Show, Hunk(h)) => editor(h.new_file.clone(), Some(h.first_diff_line())),
        (Stage, Ref(_)) => None,
        (Stage, File(u)) => cmd_arg(git::stage_file_cmd, u),
        (Stage, Delta(d)) => cmd_arg(git::stage_file_cmd, &d.new_file),
        (Stage, Hunk(h)) => cmd(h.format_patch().into_bytes(), git::stage_patch_cmd),
        (Unstage, Ref(_)) => None,
        (Unstage, File(_)) => None,
        (Unstage, Delta(d)) => cmd_arg(git::unstage_file_cmd, &d.new_file),
        (Unstage, Hunk(h)) => cmd(h.format_patch().into_bytes(), git::unstage_patch_cmd),
        (RebaseInteractive, Ref(r)) => subscreen_arg(git::rebase_interactive_cmd, r),
        (RebaseInteractive, _) => None,
        (CommitFixup, Ref(r)) => subscreen_arg(git::commit_fixup_cmd, r),
        (CommitFixup, _) => None,
        (RebaseAutosquash, Ref(r)) => subscreen_arg(git::rebase_autosquash_cmd, r),
        (RebaseAutosquash, _) => None,
        (Discard, Ref(_)) => None,
        (Discard, File(f)) => Some(Box::new(|_term, state| {
            let path = PathBuf::from_iter([state.config.dir.to_path_buf(), f.clone().into()]);
            std::fs::remove_file(path)?;
            state.screen_mut().update()
        })),
        (Discard, Delta(d)) => {
            if d.old_file == d.new_file {
                cmd_arg(git::checkout_file_cmd, &d.old_file)
            } else {
                // TODO Discard file move
                None
            }
        }
        (Discard, Hunk(h)) => cmd(
            h.format_patch().into_bytes(),
            git::discard_unstaged_patch_cmd,
        ),
        (Checkout, Ref(r)) => cmd_arg(git::checkout_ref_cmd, r),
        (Checkout, _) => None,
    }
}

fn goto_show_screen<'a, B: Backend>(r: String) -> Option<OpClosure<'a, B>> {
    Some(Box::new(move |terminal, state| {
        state.screens.push(
            screen::show::create(Rc::clone(&state.repo), terminal.size()?, r.clone())
                .expect("Couldn't create screen"),
        );
        Ok(())
    }))
}

fn editor<'a, B: Backend>(file: String, line: Option<u32>) -> Option<OpClosure<'a, B>> {
    Some(Box::new(move |terminal, state| {
        let file: &str = &file;
        let editor = std::env::var("EDITOR").expect("EDITOR not set");
        let mut cmd = Command::new(editor.clone());
        let args = match line {
            Some(line) => match editor.as_str() {
                "vi" | "vim" | "nvim" | "nano" => {
                    vec![format!("+{}", line), file.to_string()]
                }
                _ => vec![format!("{}:{}", file, line)],
            },
            None => vec![file.to_string()],
        };

        cmd.args(args);

        state
            .issue_subscreen_command(terminal, cmd)
            .expect("Error opening editor");

        state.screen_mut().update()
    }))
}

fn cmd<'a, B: Backend>(input: Vec<u8>, command: fn() -> Command) -> Option<OpClosure<'a, B>> {
    Some(Box::new(move |terminal, state| {
        state.run_external_cmd(terminal, &input, command())
    }))
}

fn cmd_arg<B: Backend>(command: fn(&str) -> Command, arg: &str) -> Option<OpClosure<B>> {
    Some(Box::new(move |terminal, state| {
        state.run_external_cmd(terminal, &[], command(arg))
    }))
}

fn subscreen_arg<B: Backend>(command: fn(&str) -> Command, arg: &str) -> Option<OpClosure<B>> {
    Some(Box::new(move |terminal, state| {
        state.issue_subscreen_command(terminal, command(arg))
    }))
}

fn goto_log_screen(repo: Rc<Repository>, screens: &mut Vec<Screen>) {
    screens.drain(1..);
    let size = screens.last().unwrap().size;
    screens.push(screen::log::create(repo, size).expect("Couldn't create screen"));
}

fn goto_refs_screen(config: &Config, screens: &mut Vec<Screen>) {
    screens.drain(1..);
    let size = screens.last().unwrap().size;
    screens.push(screen::show_refs::create(config, size).expect("Couldn't create screen"));
}

#[cfg(test)]
#[serial_test::serial]
mod tests {
    use crate::{cli::Args, update, State};
    use crossterm::event::{Event, KeyCode, KeyEvent, KeyModifiers};
    use git2::Repository;
    use ratatui::{backend::TestBackend, prelude::Rect, Terminal};
    use std::{env, fs, process::Command};
    use temp_dir::TempDir;

    #[test]
    fn no_repo() {
        let (ref mut terminal, state, _dir, _remote_dir) = setup(60, 20);
        insta::assert_snapshot!(redact_hashes(terminal, &state.repo));
    }

    #[test]
    fn help_menu() {
        let (ref mut terminal, ref mut state, _dir, _remote_dir) = setup(60, 20);
        update(terminal, state, &[key('h')]).unwrap();
        insta::assert_snapshot!(redact_hashes(terminal, &state.repo));
    }

    #[test]
    fn fresh_init() {
        let (ref mut terminal, ref mut state, _dir, _remote_dir) = setup(60, 20);
        update(terminal, state, &[key('g')]).unwrap();
        insta::assert_snapshot!(redact_hashes(terminal, &state.repo));
    }

    #[test]
    fn new_file() {
        let (ref mut terminal, ref mut state, dir, _remote_dir) = setup(60, 20);
        run(&dir, &["touch", "new-file"]);
        update(terminal, state, &[key('g')]).unwrap();
        insta::assert_snapshot!(redact_hashes(terminal, &state.repo));
    }

    #[test]
    fn unstaged_changes() {
        let (ref mut terminal, ref mut state, dir, _remote_dir) = setup(60, 20);
        commit(&dir, "testfile", "testing\ntesttest");
        fs::write(dir.child("testfile"), "test\ntesttest").expect("error writing to file");

        update(terminal, state, &[key('g')]).unwrap();
        insta::assert_snapshot!(redact_hashes(terminal, &state.repo));
    }

    #[test]
    fn staged_file() {
        let (ref mut terminal, ref mut state, dir, _remote_dir) = setup(60, 20);
        run(&dir, &["touch", "new-file"]);
        run(&dir, &["git", "add", "new-file"]);
        update(terminal, state, &[key('g')]).unwrap();
        insta::assert_snapshot!(redact_hashes(terminal, &state.repo));
    }

    #[test]
    fn log() {
        let (ref mut terminal, ref mut state, dir, _remote_dir) = setup(60, 20);
        commit(&dir, "firstfile", "testing\ntesttest\n");
        commit(&dir, "secondfile", "testing\ntesttest\n");
        update(terminal, state, &[key('g'), key('l'), key('l')]).unwrap();
        insta::assert_snapshot!(redact_hashes(terminal, &state.repo));
    }

    #[test]
    fn show() {
        let (ref mut terminal, ref mut state, dir, _remote_dir) = setup(60, 20);
        commit(&dir, "firstfile", "This should be visible\n");
        update(
            terminal,
            state,
            &[key('g'), key('l'), key('l'), key_code(KeyCode::Enter)],
        )
        .unwrap();
        insta::assert_snapshot!(redact_hashes(terminal, &state.repo));
    }

    #[test]
    fn rebase_conflict() {
        let (ref mut terminal, ref mut state, dir, _remote_dir) = setup(60, 20);
        commit(&dir, "new-file", "hello");

        run(&dir, &["git", "checkout", "-b", "other-branch"]);
        commit(&dir, "new-file", "hey");

        run(&dir, &["git", "checkout", "main"]);
        commit(&dir, "new-file", "hi");

        run(&dir, &["git", "checkout", "other-branch"]);
        run(&dir, &["git", "rebase", "main"]);

        update(terminal, state, &[key('g')]).unwrap();
        insta::assert_snapshot!(redact_hashes(terminal, &state.repo));
    }

    #[test]
    fn merge_conflict() {
        let (ref mut terminal, ref mut state, dir, _remote_dir) = setup(60, 20);
        commit(&dir, "new-file", "hello");

        run(&dir, &["git", "checkout", "-b", "other-branch"]);
        commit(&dir, "new-file", "hey");

        run(&dir, &["git", "checkout", "main"]);
        commit(&dir, "new-file", "hi");

        run(&dir, &["git", "merge", "other-branch"]);

        update(terminal, state, &[key('g')]).unwrap();
        insta::assert_snapshot!(redact_hashes(terminal, &state.repo));
    }

    #[test]
    fn moved_file() {
        let (ref mut terminal, ref mut state, dir, _remote_dir) = setup(60, 20);
        commit(&dir, "new-file", "hello");
        run(&dir, &["git", "mv", "new-file", "moved-file"]);

        update(terminal, state, &[key('g')]).unwrap();
        insta::assert_snapshot!(redact_hashes(terminal, &state.repo));
    }

    #[test]
    fn hide_untracked() {
        let (ref mut terminal, ref mut state, dir, _remote_dir) = setup(60, 10);
        let mut config = state.repo.config().unwrap();
        config.set_str("status.showUntrackedFiles", "off").unwrap();
        run(&dir, &["touch", "i-am-untracked"]);

        update(terminal, state, &[key('g')]).unwrap();
        insta::assert_snapshot!(redact_hashes(terminal, &state.repo));
    }

    #[test]
    fn push() {
        let (ref mut terminal2, ref mut state2, dir2, _remote_dir) = setup(60, 10);
        commit(&dir2, "second-file", "");

        update(terminal2, state2, &[key('P'), key('p')]).unwrap();
        insta::assert_snapshot!(redact_hashes(terminal2, &state2.repo));
    }

    fn setup(width: u16, height: u16) -> (Terminal<TestBackend>, State, TempDir, TempDir) {
        let terminal = Terminal::new(TestBackend::new(width, height)).unwrap();
        let remote_dir = TempDir::new().unwrap();
        let dir = TempDir::new().unwrap();

        Repository::init_bare(remote_dir.path()).unwrap();
        Repository::clone(remote_dir.path().to_str().unwrap(), dir.path()).unwrap();

        let state = State::create(
            Repository::open(dir.path()).unwrap(),
            crate::Config {
                dir: dir.path().into(),
            },
            Rect::new(0, 0, width, height),
            Args {
                command: None,
                status: false,
                exit_immediately: false,
            },
        )
        .unwrap();

        let mut config = state.repo.config().unwrap();
        config.set_str("user.email", "ci@example.com").unwrap();
        config.set_str("user.name", "CI").unwrap();

        (terminal, state, dir, remote_dir)
    }

    fn commit(dir: &TempDir, file_name: &str, contents: &str) {
        let path = dir.child(file_name);
        let message = match path.try_exists() {
            Ok(true) => format!("modify {}\n\nCommit body goes here\n", file_name),
            _ => format!("add {}\n\nCommit body goes here\n", file_name),
        };
        fs::write(path, contents).expect("error writing to file");
        run(dir, &["git", "add", file_name]);
        run(dir, &["git", "commit", "-m", &message]);
    }

    fn run(dir: &TempDir, cmd: &[&str]) {
        Command::new(cmd[0])
            .args(&cmd[1..])
            .env("GIT_COMMITTER_DATE", "Sun Feb 18 14:00 2024 +0100")
            .current_dir(dir.path())
            .output()
            .unwrap_or_else(|_| panic!("failed to execute {:?}", cmd));
    }

    fn key(char: char) -> Event {
        let mods = if char.is_uppercase() {
            KeyModifiers::SHIFT
        } else {
            KeyModifiers::empty()
        };

        Event::Key(KeyEvent::new(KeyCode::Char(char), mods))
    }

    fn key_code(code: KeyCode) -> Event {
        Event::Key(KeyEvent::new(code, KeyModifiers::empty()))
    }

    fn redact_hashes(terminal: &mut Terminal<TestBackend>, repo: &Repository) -> String {
        let mut debug_output = format!("{:#?}", terminal.backend().buffer());

        let mut revwalk = repo.revwalk().unwrap();
        revwalk.push_head().ok();
        revwalk
            .flat_map(|maybe_oid| maybe_oid.and_then(|oid| repo.find_commit(oid)))
            .for_each(|commit| {
                let id = commit.as_object().id().to_string();
                let short = commit.as_object().short_id().unwrap();
                let short_id = short.as_str().unwrap();

                debug_output = debug_output.replace(&id, &"_".repeat(id.len()));
                debug_output = debug_output.replace(short_id, &"_".repeat(short_id.len()));
            });

        debug_output
    }
}
