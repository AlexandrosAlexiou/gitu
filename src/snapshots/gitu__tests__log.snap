---
source: src/lib.rs
expression: "redact_hashes(terminal, &state.repo)"
---
Buffer {
    area: Rect { x: 0, y: 0, width: 60, height: 20 },
    content: [
        "ðŸ¢’_______ main add secondfile                                ",
        " _______ add firstfile                                      ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
    ],
    styles: [
        x: 0, y: 0, fg: Reset, bg: Rgb(80, 73, 69), underline: Reset, modifier: NONE,
        x: 1, y: 0, fg: Rgb(216, 166, 87), bg: Rgb(80, 73, 69), underline: Reset, modifier: NONE,
        x: 8, y: 0, fg: Reset, bg: Rgb(80, 73, 69), underline: Reset, modifier: NONE,
        x: 9, y: 0, fg: Rgb(169, 182, 101), bg: Rgb(80, 73, 69), underline: Reset, modifier: NONE,
        x: 13, y: 0, fg: Reset, bg: Rgb(80, 73, 69), underline: Reset, modifier: NONE,
        x: 0, y: 1, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 1, fg: Rgb(216, 166, 87), bg: Reset, underline: Reset, modifier: NONE,
        x: 8, y: 1, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
    ]
}
